docs/_tooling/extensions/header_service/test/test_header_service.py

```diff
@@ -0,0 +1,321 @@
+@patch("sphinx.application.Sphinx")
+def test_request_from_directive_github_data(
+    mock_app, mock_extract_github_data, mock_generate_hash
+):
+    github_data = {
+        "author": "John Doe",
+        "approvers": "apover_1",
+        "reviewers": "reviewer_1",
+        "hash": "abcdef",
+    }
+    hash_value = "abcdef"
+    need_data = {
+        "type": "review_header",
+        "title": "Service Title",
+        "layout": "focus",
+        "template": "header",
+        "id": f"RH__{hash_value}",
+    }
+    mock_extract_github_data.return_value = github_data
+    mock_generate_hash.return_value = hash_value
+    mock_directive = MagicMock()
+    mock_directive.env.config.header_service_use_github_data = True
+    header_service = hs.HeaderService(mock_app, "dummy-service", None)
+    assert header_service.request_from_directive(mock_directive) == [
+        need_data | github_data
+    ]
+
+
+@patch("header_service.header_service.generate_hash")
+@patch("header_service.header_service._extract_merge_commit_data")
+@patch("sphinx.application.Sphinx")
+def test_request_from_directive_commit_data(
+    mock_app, mock_extract_merge_commit_data, mock_generate_hash
+):
+    github_data = {
+        "author": "John Doe",
+        "approvers": "apover_1",
+        "reviewers": "reviewer_1",
+        "hash": "abcdef",
+    }
+    hash_value = "abcdef"
+    need_data = {
+        "type": "review_header",
+        "title": "Service Title",
+        "layout": "focus",
+        "template": "header",
+        "id": f"RH__{hash_value}",
+    }
+    mock_directive = MagicMock()
+    mock_directive.env.config.header_service_use_github_data = False
+    mock_directive.env.docname = "file1.rst"
+    mock_extract_merge_commit_data.return_value = github_data
+    mock_generate_hash.return_value = hash_value
+    header_service = hs.HeaderService(
+        mock_app, "dummy-service", {"header_service_use_github_data": False}
+    )
+    assert header_service.request_from_directive(mock_directive) == [
+        need_data | github_data
+    ]
+    mock_extract_merge_commit_data.assert_called_once_with("file1.rst")
+
+
+@patch("header_service.header_service.HeaderService.request_from_directive")
+@patch("sphinx.application.Sphinx")
+def test_debug(mock_app, mock_request_from_directive):
+    debug_data = [{"key": "value"}]
+    mock_request_from_directive.return_value = debug_data
+    header_service = hs.HeaderService(mock_app, "dummy-service", None)
+    assert header_service.debug(None) == debug_data
+
+
+@patch("header_service.header_service.subprocess.run")
+def test_extract_merge_commit_data(run_mock):
+    lines = """abcdef
+John Doe
+Approved: {approver1} ( {approver1@mail.com} ) on {2024-11-1}
+Approved: {approver2} ( {approver2@mail.com} ) on {2024-11-2}
+Reviewed: {reviewer1} ( {reviewer1@mail.com} ) on {2024-12-1}
+Reviewed: {reviewer2} ( {reviewer2@mail.com} ) on {2024-12-2}
+Reviewed: {reviewer3} ( {reviewer3@mail.com} ) on {2024-12-3}"""
+
```

Comment:
The tests need some work on readability, although I'm not sure yet how.

An example here would be something like: `run_mock = set_stdout_for_next_run_call(lines)`.

Reply:
I was thinking about a contextmanager as well. Not sure about fixtures, I have no experience with mocking.